File: ros2_ws/src/exploration_robot_exploration/src/planning/track.cpp
#include "exploration_robot_exploration/planning/track.hpp"

#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace exploration_robot_exploration
{

namespace
{
constexpr double kPi = 3.14159265358979323846;
double wrap_angle(const double angle) {return std::remainder(angle, 2.0 * kPi);}
}  // namespace

Track::Track(
  exploration_robot_control::AccelerationController & controller,
  const double waypoint_tolerance,
  const double target_speed,
  const double linear_gain,
  const double angular_gain,
  const double angular_damping)
: controller_(controller),
  waypoint_tolerance_(waypoint_tolerance),
  target_speed_(target_speed),
  linear_gain_(linear_gain),
  angular_gain_(angular_gain),
  angular_damping_(angular_damping)
{
  if (waypoint_tolerance_ <= 0.0 || target_speed_ <= 0.0) {
    throw std::invalid_argument("Tolerance and target speed must be positive");
  }
}

const Waypoints & Track::create_route(const Coordinate & start, const Coordinate & goal)
{
  waypoints_.replace(plan_route(start, goal));
  return waypoints_;
}

bool Track::follow_waypoint(const exploration_robot_control::RobotState & state)
{
  auto target = waypoints_.current();
  if (!target) {
    controller_.reset();
    return true;
  }

  double dx = (*target)[0] - state.x;
  double dy = (*target)[1] - state.y;
  double distance = std::hypot(dx, dy);
  if (distance <= waypoint_tolerance_) {
    waypoints_.advance();
    target = waypoints_.current();
    if (!target) {
      controller_.reset();
      return true;
    }
    dx = (*target)[0] - state.x;
    dy = (*target)[1] - state.y;
    distance = std::hypot(dx, dy);
  }

  const double heading_error = wrap_angle(std::atan2(dy, dx) - state.theta);
  const double alignment = std::max(0.0, 1.0 - std::abs(heading_error) / (kPi / 2.0));
  const double desired_speed = std::min(target_speed_, distance) * alignment;
  const double linear_acceleration = linear_gain_ * (desired_speed - state.linear_velocity);
  const double angular_acceleration =
    angular_gain_ * heading_error - angular_damping_ * state.angular_velocity;
  controller_.apply_control(linear_acceleration, angular_acceleration);
  return false;
}

bool Track::route_complete() const noexcept {return waypoints_.complete();}
const Waypoints & Track::waypoints() const noexcept {return waypoints_;}
exploration_robot_control::AccelerationController & Track::controller() noexcept
{
  return controller_;
}

}  // namespace exploration_robot_exploration

File: Logic/Robot/Track.py
"""Planificación abstracta y seguimiento de rutas para el robot."""

from __future__ import annotations

from abc import ABC, abstractmethod
from math import atan2, hypot, pi
from typing import Sequence

from .Physic import RobotPhysics
from .waypoints import Waypoints

Coordinate = Sequence[float]
CoordinateMatrix = list[list[float]]


def _wrap_angle(angle: float) -> float:
    return (angle + pi) % (2.0 * pi) - pi


class Track(ABC):
    def __init__(
        self,
        robot: RobotPhysics,
        *,
        waypoint_tolerance: float = 0.10,
        target_speed: float = 1.0,
        linear_gain: float = 1.5,
        angular_gain: float = 4.0,
        angular_damping: float = 1.5,
    ) -> None:
        if not isinstance(robot, RobotPhysics):
            raise TypeError("robot debe ser una instancia de RobotPhysics")
        if waypoint_tolerance <= 0.0:
            raise ValueError("waypoint_tolerance debe ser mayor que cero")
        if target_speed <= 0.0:
            raise ValueError("target_speed debe ser mayor que cero")
        self.robot = robot
        self.waypoints = Waypoints()
        self.waypoint_tolerance = float(waypoint_tolerance)
        self.target_speed = float(target_speed)
        self.linear_gain = float(linear_gain)
        self.angular_gain = float(angular_gain)
        self.angular_damping = float(angular_damping)

    @abstractmethod
    def plan_route(
        self, start: Coordinate, goal: Coordinate
    ) -> CoordinateMatrix:
        """Calcula y devuelve una matriz de coordenadas [[x, y], ...]."""
        raise NotImplementedError

    def create_route(self, start: Coordinate, goal: Coordinate) -> Waypoints:
        self.waypoints.replace(self.plan_route(start, goal))
        return self.waypoints

    def compute_control(self) -> tuple[float, float] | None:
        """Calcula el control nominal hacia el waypoint activo.

        Devuelve ``None`` cuando la ruta ha terminado. Este método no modifica
        las físicas; el filtro SafeTracker se ejecuta antes de ``apply_control``.
        """
        target = self.waypoints.current
        if target is None:
            return None

        state = self.robot.state
        dx, dy = target[0] - state.x, target[1] - state.y
        distance = hypot(dx, dy)
        if distance <= self.waypoint_tolerance:
            self.waypoints.advance()
            target = self.waypoints.current
            if target is None:
                return None
            dx, dy = target[0] - state.x, target[1] - state.y
            distance = hypot(dx, dy)

        desired_heading = atan2(dy, dx)
        heading_error = _wrap_angle(desired_heading - state.theta)
        alignment = max(0.0, 1.0 - abs(heading_error) / (pi / 2.0))
        desired_speed = min(self.target_speed, distance) * alignment
        linear_acceleration = self.linear_gain * (
            desired_speed - state.linear_velocity
        )
        angular_acceleration = (
            self.angular_gain * heading_error
            - self.angular_damping * state.angular_velocity
        )
        return linear_acceleration, angular_acceleration

    def follow_waypoint(self) -> bool:
        """Compatibilidad para seguimiento sin filtro de seguridad."""
        nominal_control = self.compute_control()
        if nominal_control is None:
            self.robot.stop()
            return True
        self.robot.apply_control(*nominal_control)
        return False

    @property
    def route_complete(self) -> bool:
        return self.waypoints.complete
